### PR TITLE
8 Metrics and graphics under menu button Bürokratt

### PR DESCRIPTION
**Pull request by baha-a**
Add DSL/Resql files and GUI translation files.

~~The [related task](https://github.com/buerokratt/Analytics-Module/issues/8) is blocked now until both GUI design is ready and `intents` metric data is available.~~

### CODE REVIEW

File: DSL/Ruuter/POST/avg-session-length-client-left.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+extractRequestData:
```
Comment (1AhmedYasser) [inline DSL/Ruuter/POST/avg-session-length-client-left.yml:1]: Add Parameter validation here as well

File: DSL/Ruuter/POST/avg-session-length-csa.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+extractRequestData:
```
Comment (1AhmedYasser) [inline DSL/Ruuter/POST/avg-session-length-csa.yml:1]: Add Parameter validation here as well

File: DSL/Ruuter/POST/avg-session-length-no-csa.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+extractRequestData:
```
Comment (1AhmedYasser) [inline DSL/Ruuter/POST/avg-session-length-no-csa.yml:1]: Add Parameter validation here as well

File: DSL/Ruuter/POST/pct-currectly-understood.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+extractRequestData:
```
Comment (1AhmedYasser) [inline DSL/Ruuter/POST/pct-currectly-understood.yml:1]: Add Parameter validation here as well

**Review by 1AhmedYasser — DISMISSED** [review]
Add Parameters Validation

**Review by gdrui — CHANGES_REQUESTED** [review]
The DSL here is not what is required by the issue, the query should be composed based on the metric, additional options and time period and the result should be a dataset grouped by day or hour, so it could be rendered as a chart. 

Also, lets start keeping Ruuter yml files in the folder that reflects which tab they belong to, in this case `buerokratt`.

Another note is that you could probably have one Ruuter endpoint to handle all the different options and have that retrieve the data from a specific Resql endpoint.

**Comment by 1AhmedYasser** [comment]
> Also, lets start keeping Ruuter yml files in the folder that reflects which tab they belong to, in this case `buerokratt`.

Yup Should Also modify the dsl to handle all 5 period options as follows for example

```
SELECT date_trunc(:metric, created) AS date_time,
       COUNT(DISTINCT base_id)
FROM chat
WHERE EXISTS
    (SELECT 1
     FROM message
     WHERE message.chat_base_id = chat.base_id
       AND message.author_role = 'backoffice-user')
  AND EXISTS
    (SELECT 1
     FROM message
     WHERE message.chat_base_id = chat.base_id
       AND message.author_role = 'end-user')
  AND created::date BETWEEN :start::date AND :end::date
GROUP BY 1
```

File: DSL/Resql/byk-avg-response-speed.sql [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        m.chat_base_id,
+        (
+            SELECT MIN(created) 
+            FROM message m2 
+            WHERE m2.author_role <> 'chatbot' 
+            AND m2.created < m.created 
+            AND m2.chat_base_id = m.chat_base_id
+        ) AS previous_message_time
+    FROM message m
+    JOIN chat c ON m.chat_base_id = c.base_id
```
Comment (gdrui) [inline DSL/Resql/byk-avg-response-speed.sql:14]: you dont actually use `chat` in this query

File: DSL/Resql/byk-avg-response-speed.sql [hunk 6]
```diff
@@ -0,0 +1,30 @@
+WITH chatbot_messages AS (
+    SELECT 
+        m.author_role, 
+        m.created,
+        m.chat_base_id,
+        (
+            SELECT MIN(created)
```
Comment (gdrui) [inline DSL/Resql/byk-avg-response-speed.sql:7]: hmm, don't you mean the `MAX(created)` to get the latest message before the chat's message?

File: DSL/Resql/byk-avg-sessions-time-customer.sql [hunk 7]
```diff
@@ -0,0 +1,17 @@
+WITH closed_chats AS (
+    SELECT
+        chat_base_id,
+        MIN(m.created) AS start_time,
+        MAX(m.created) AS end_time
+    FROM message m
+    JOIN chat ON m.chat_base_id = chat.base_id
+    WHERE chat.status IN ('ANSWERED', 'CLIENT-LEFT')
```
Comment (gdrui) [inline DSL/Resql/byk-avg-sessions-time-customer.sql:8]: AFAIK these are not values in chat.status but values in message.event, so these should be chats that have any messages with these events.

File: DSL/Ruuter/GET/buerokratt/byk-avg-sessions-time.yml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  next: check_byk
+
+missing_parameter:
+  status: 400
+  return: 'required parameters were not provided'
+  next: end
+
+check_byk:
+  switch:
+    - condition: ${options.includes('byk')}
```
Comment (gdrui) [inline DSL/Ruuter/GET/buerokratt/byk-avg-sessions-time.yml:26]: this is a clever way to combine multiple options, even though Ruuter makes the readability a little difficult

**Review by gdrui — CHANGES_REQUESTED** [review]

File: DSL/Resql/byk-avg-response-speed.sql (hunk 6 above)
Comment (baha-a) [reply DSL/Resql/byk-avg-response-speed.sql:7]: You are right, it should be `MAX`.
Fixed.

File: DSL/Resql/byk-avg-sessions-time-customer.sql (hunk 7 above)
Comment (baha-a) [reply DSL/Resql/byk-avg-sessions-time-customer.sql]: Fixed

File: DSL/Resql/byk-avg-response-speed.sql (hunk 5 above)
Comment (baha-a) [reply DSL/Resql/byk-avg-response-speed.sql]: Fixed

**Comment by baha-a** [comment]
For the intents metrics, I assumed that there is a table called `intents` connected with table `message` by `message.intent_base_id`.

**Review by gdrui — APPROVED** [review]

**Pull request closed without merging**